### ISSUE TITLE
Memory leak with the consume(n) method and limit in consumeLoop method

### ISSUE THREAD

**Issue by AAraKKe** [Blizzard/node-rdkafka#720]
Hi,

we have been noticing an issue with how our consumers behave since some time ago and we cannot figure what the issue might be. The only explanation is something in the internals of the library that escapes our understanding and we cannot manage properly. Let me explain:

We have a series of services, each of them defining a consumer and a producer. These services will basically read a topic, used a message pipe, react to them and produce back a message when it has finished. The way we are consuming messages is with the following configuration:
```ts
        this.consumerConfig = {
            "group.id": groupName,
            "metadata.broker.list": brokerList,
            "enable.auto.commit": false,
            "enable.auto.offset.store": false,
            "max.poll.interval.ms": maxPollMs
        };
        this.consumerTopicConfig = {
            "enable.auto.commit": false,
            "auto.offset.reset": offsetReset
        };
        this.consumer = new KafkaConsumer(this.consumerConfig, this.consumerTopicConfig);
        // Later in the class
        this.consumer.on('ready', () => {
            this.consumer.subscribe(topicNames);
            this.streamTimeout = setInterval(() => {
                this.consumer.consume(1);
            }, 1000);
        })
```

What we do later is to react to data events to handle the message. We observed a memory leak since September 9th although this date might not be important since the environment is set in docker and a new release had not been done since August. Without any changes on the consumer itself, or anything related with the consumer, memory usage starts to increase without being cleaned by the garbage collector.

I am not sharing what we do when consume data because this behaviour is found even when leaving the service running with no messages. It does not matter if there are messages to handle or not, the fact that the consumer is consuming makes memory usage go up with no limit. We narrowed it down to this part of the code and if we just call that method once, then no memory leak appears and everything behaves properly.

We found that if we use the consumeLoop call (by calling consume with no arguments) then the memory issue disapears as well. This is

```ts
            this.consumer.setDefaultConsumeTimeout(1000);
            this.consumer.consume();
```

Here we can consume in the same way and memory behaviour is way better. Nonetheless there is an issue with this. When we launch the entire suite that can include up to 9 services (running in the same container) only 4 of them consume data, like if there was a limit to the number of consumers that can consume data in this way.

Each consumer is independent but all of them are running in the same node app. If we split each service in different containers, running 9 node apps each of them with one consumer then the issue disappears and all consumers read properly.

This is our end goal but given the limitations we have now it is not easy for use to switch the infrastructure in the short term and I would like to hear your opinion about this.

Why do you think the memory leaks happens and if is there any limit in the consumers that can run simultaneously with the consume() method. I have not been able to answer any of these questions so far and would love some input about it.

Thanks!

**Comment by Tapppi** [comment]
> What we do later is to react to data events to handle the message.

Why do you want to  `.consume(size, cb)` if you are anyway reacting to `data` events in a separate handler? Not saying it is wrong, just interesting.

> Then we launch the entire suite that can include up to 9 services (running in the same container) only 4 of them consume data, like if there was a limit to the number of consumers that can consume data in this way.

Each kafka consumer needs its own `libuv` worker thread to fetch messages, and the default `libuv` thread count is only 4. That is why the rest of your consumers are not working. You can increase the thread count by setting the `UV_THREADPOOL_SIZE` environment variable.

**Comment by AAraKKe** [comment]
Thanks for the answer @Tapppi ! We do not use the callback because of how the code is structured. A part of the class defines the parts that execute and another part of the class defines the answers to events, not only kafka launches the events. We could do it the other way around, but it was just a design choice.

Thanks for pointing out the environment variable, I was not aware of that and should fix the issue. Is it safe to increase this limit or could there be side effects related with the implementation?

Nonetheless, it would be great to know why using the `consume(size)` method is causing a memory leak. It was ok until a new release we did in November. This release changed nothing in how we consume messages, i.e. it was unrelated to the kafka part of the project and the memory leak appeared. Not sure if there was some underlying change in the librdkafka that was picked up, although I imagine the librdkafka version is frozen in the node one. Do you have a feeling where the memory leak is coming from and have you seen this before?

**Comment by codeburke** [comment]
I think we need some more information:
1. What version of node-rdkafka are you using? 
2. What version of Node are you using?
3. Did you update either the Node or node-rdkafka versions in the November release?

I can say that I am using the `consume(N)` API at scale at my work with no memory leaks using the latest Node 12 LTS and node-rdkafka 2.7.4.

**Comment by AAraKKe** [comment]
The node-rdkafka version is frozen in the 2.7.0 and the reason is that above that version producers do not send messages when running in AWS with MSK. I updated node-rdkafka to the latest version to see if that would fix the memory leak and pushed to our staging environment in AWS. Our producers that send messages when the services finish did not produce any message to MSK but they worked in a kafka container locally so there might be some issue with it in AWS. The last version that worked is 2.7.0 and this why the version is frozen. If you want I can open a separate issue for that so it can be investigated. We just ignored it for now because MSK is rather new but it is probably good if you could take a look at that.

Node is 11.15.0, that is the version we have configured in our docker image and neither node nor node-rdkafka have been modified since we updated to 2.7.0 in July. The updates pushed to production in November were not related with the event handling part of the project, only with the services that run when a message is received which are independent from the event handling part of the code. The memory leak happens even when we do not send or receive messages, just letting the consumer run with the `consume(N)` is enough to trigger the memory issue.

**Comment by Tapppi** [comment]
> Thanks for pointing out the environment variable, I was not aware of that and should fix the issue. Is it safe to increase this limit or could there be side effects related with the implementation?

It needs to be better documented, this comes up regularly in issues.. It is "safe" as long as the process can handle the thread count without resource starvation (CPU, memory allocation). It just allows `libuv` to spawn more threads for work by different consumers and having `UV_THREADPOOL_COUNT > consumerCount` means that locks/sleeps/etc on the worker threads won't block other consumers.

> Why does it not raise an exception if more consumers are added?

It isn't an exception really, just means there will be resource contention between the consumers. I guess the library could warn about this though. Even with `UV_THREADPOOL_COUNT < consumerCount` you should eventually see all consumers progress but there can be very long pauses (or indefinite if some consumers never reach the end of the topic) because of other consumers hogging the thread. But the library doesn't handle that situation particularly well so not recommended, we had regular 15 minute delay when we forgot to increase this variable with consumer count at work..

> so there might be some issue with it in AWS. The last version that worked is 2.7.0 and this why the version is frozen.

You should open another issue for that.

> Node is 11.15.0, that is the version we have configured in our docker image and neither node nor node-rdkafka have been modified since we updated to 2.7.0 in July

Just to note, it is recommended to use LTS versions in production unless you have a good reason not to. Node 11 has been unsupported since the release of Node 12 6 month ago, so it was already unsupported when you changed to it. See [Node release and support schedule](https://nodejs.org/en/about/releases/).

> just letting the consumer run with the consume(N) is enough to trigger the memory issue

If you can make a small repo with code that reproduces this, it will be a lot easier to debug. That said, I don't have time to dig in right now, hopefully someone else will..

**Comment by codeburke** [comment]
> If you can make a small repo with code that reproduces this, it will be a lot easier to debug.
+1, that would be very helpful in debugging if you can create a simple reproducible repo for us to help debug on.

**Comment by AAraKKe** [comment]
Hi, thanks for the quick answers and sorry for the delay. We started using node 11 when it was still maintained and node 12 did not come out yet (it was end of 2018) and since then we have just never reivewed this. I trie dupdating to node 12.x and node-rdkafka actually failed compiling.

Using the following docker file:

```
FROM python:3.7 as base
RUN curl -sL https://deb.nodesource.com/setup_12.x | bash - && \
    apt-get update && apt-get install -y --no-install-recommends nodejs && \
    rm -rf /var/lib/apt/lists/* && \
    node -v && npm -v
WORKDIR /app
COPY package.json /app
RUN npm install
```

where package.json contains `"node-rdkafka": "^2.7.0"` as dependency, sould give you the same issue. In this case we are using node v12.14.0 and npm 6.13.4, these are the versions installed.

The error we get is the following:

```
Traceback (most recent call last):
  File "/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py", line 50, in <module>
    sys.exit(gyp.script_main())
  File "/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 554, in script_main
    return main(sys.argv[1:])
  File "/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 547, in main
    return gyp_main(args)
  File "/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/pylib/gyp/__init__.py", line 532, in gyp_main
    generator.GenerateOutput(flat_list, targets, data, params)
  File "/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/pylib/gyp/generator/make.py", line 2215, in GenerateOutput
    part_of_all=qualified_target in needed_targets)
  File "/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/pylib/gyp/generator/make.py", line 794, in Write
    extra_mac_bundle_resources, part_of_all)
  File "/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/pylib/gyp/generator/make.py", line 978, in WriteActions
    part_of_all=part_of_all, command=name)
  File "/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/pylib/gyp/generator/make.py", line 1724, in WriteDoCmd
    force = True)
  File "/usr/lib/node_modules/npm/node_modules/node-gyp/gyp/pylib/gyp/generator/make.py", line 1779, in WriteMakeRule
    cmddigest = hashlib.sha1(command if command else self.target).hexdigest()
TypeError: Unicode-objects must be encoded before hashing
gyp ERR! configure error
gyp ERR! stack Error: `gyp` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onCpExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/configure.js:351:16)
gyp ERR! stack     at ChildProcess.emit (events.js:210:5)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12)
gyp ERR! System Linux 4.9.184-linuxkit
gyp ERR! command "/usr/bin/node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /app/node_modules/node-rdkafka
gyp ERR! node -v v12.14.0
gyp ERR! node-gyp -v v5.0.5
gyp ERR! not ok
npm WARN customer360-graph-builder@1.0.0 license should be a valid SPDX license expression

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! node-rdkafka@2.7.4 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the node-rdkafka@2.7.4 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2019-12-18T08_08_12_199Z-debug.log
```

**Comment by skairali** [comment]
Any update on this issue?  we are also seeing memory leak

**Comment by skairali** [comment]
@AAraKKe  did you use UV_THREADPOOL_COUNT?  DId it improve the memory leak situation?

**Comment by AAraKKe** [comment]
Sorry for the delay! We have been very busy with other parts of the project and this kind of slipped out of my mind. Yes, the UV_THREADPOOL_COUNT fixed the issue. We can now have more consumers reading. For the memory leak we are still using the workaround I mentioned of used the consumeLoop method and leave it to the package to handle it. With the env variable change it is working fine.

I noted down the update to node 12 although I am not sure how fast I will be able to give you any update on that. We still cannot update to the latest version of the package because doing so prevents producers from sending messages when running in AWS and MSK, as mentioned before. 

As soon as I can I will try updating node, this package and see if that fixes the issue because maybe is some kind of incompatibility I cannot report now. Thanks.

**Comment by fuqiang2310** [comment]
@AAraKKe I recently encountered the same memory issue with the consume(n) method. After digging into it I found out the way how this method work in librdkafka is, whenever this function is called, it will create a new worker to poll messages from kafka broker. The worker will not die until it polls n number of messages. Therefore in your code:
```
this.streamTimeout = setInterval(() => {
                this.consumer.consume(1);
            }, 1000);
```
When there is no new message, you are still creating new workers to poll 1 message every second. Adding a flag to stop this will be an easy solution:
```
setInterval(function() {
          if (!isConsuming) {
            isConsuming = true;
            kafkaConsumer.consume(1);
          }
        }, 1000);
```
